completions/tar

```diff
@@ -453,33 +468,33 @@ _comp_cmd_tar__detect_ext()
     fi
     ext="$tars"
 
-    case "$tar_mode_arg" in
-        --*)
-            # Should never happen?
-            ;;
-        ?(-)*[cr]*f)
+    #TODO: lzip, lzma, lzop
+    if ! _comp_cmd_tar__extract_like_mode; then
+        if ! _comp_cmd_tar__is_bsdtar; then
             ext='@(tar|gem|spkg|cbt|xpbs)'
-            case ${words[1]} in
-                *a*) ext="$tars" ;;
-                *z*) ext='t?(ar.)gz' ;;
-                *Z*) ext='ta@(r.Z|z)' ;;
-                *[jy]*) ext='t@(?(ar.)bz?(2)|b2)' ;;
-                *J*) ext='t?(ar.)xz' ;;
-            esac
-            ;;
-        +([^ZzJjy])f)
-            # Pass through using defaults above
-            ;;
-        *[Zz]*f)
-            ext='@(@(t?(ar.)|spkg.)@(gz|Z)|taz)'
-            ;;
-        *[jy]*f)
-            ext='@(t?(ar.)bz?(2)|spkg|tb2)'
-            ;;
-        *[J]*f)
-            ext='@(@(tar|spkg).@(lzma|xz)|t[lx]z)'
-            ;;
-    esac
+        fi
+        if [[ $tar_mode_arg == *a* || $compression_mode == auto-compress ]]; then
+            ext="$tars"
+        elif [[ $tar_mode_arg == *z* || $compression_mode == gzip ]]; then
+            ext='t?(ar.)gz'
+        elif [[ $tar_mode_arg == *Z* || $compression_mode == compress ]]; then
+            ext='ta@(r.Z|z)'
+        elif [[ $tar_mode_arg == *[jy]* || $compression_mode == bzip2 ]]; then
+            ext='t@(?(ar.)bz?(2)|b2)'
+        elif [[ $tar_mode_arg == *J* || $compression_mode == xz ]]; then
+            ext='t?(ar.)xz'
+        fi
+    elif [[ $tar_mode_arg == +([^ZzJjy])f && $compression_mode == none ]]; then
+        : # Pass through using defaults above
+    elif [[ $tar_mode_arg == *[Zz]*f || $compression_mode == gzip || $compression_mode == compress ]]; then
+        ext='@(@(t?(ar.)|spkg.)@(gz|Z)|taz)'
+    elif [[ $tar_mode_arg == *[jy]*f || $compression_mode == bzip2 ]]; then
+        ext='@(t?(ar.)bz?(2)|spkg|tb2)'
+    elif [[ $tar_mode_arg == *[J]*f || $compression_mode == xz ]]; then
+        ext='@(@(tar|spkg).@(lzma|xz)|t[lx]z)'
+    elif [[ $compression_mode == zstd ]]; then
+        ext='t?(ar.)zst'
+    fi
```

Comment:
These could be more readable when turned into `case $tar_mode_arg in` blocks.

Reply:
Do you mean putting `if [[ $compression_mode ==` inside each case? Or something else?

Reply:
You could do `case $tar_mode_arg:$compression_mode in` and match it to `+([^ZzJjy])f:none` or `*[Zz]*f:* | *:gzip | *:compress`, etc.

Reply:
Yep, that's even better :+1:

Reply:
Did that, it conflicted a bit with the bsdtar change, so I put it in the same commit with the compression change.